File: app/src/main/java/com/softwareverde/tidyduck/most/MostFunction.java
package com.softwareverde.tidyduck.most;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;

/**
 * <p>Represents a MOST function.</p>
 *
 * <p>This class is abstract as it does not implement the
 * XmlNode generateXmlElement method.  This is left for
 * concrete implementations (e.g. Property and Method) to
 * implements since the XML export may change depending
 * on the function type.</p>
 */
public abstract class MostFunction {
    private Long _id;
    private String _mostId;
    private String _name;
    private String _description;
    private String _release;
    private boolean _isDeleted;
    private Date _deletedDate;
    private boolean _isPermanentlyDeleted;
    private Date _permanentlyDeletedDate;
    private boolean _isApproved;
    private boolean _isReleased;
    private Long _approvalReviewId;
    private MostFunctionStereotype _functionStereotype;
    private Author _author;
    private Company _company;
    private String _returnParameterName;
    private String _returnParameterDescription;
    private MostType _returnType;
    private List<Operation> _operations = new ArrayList<>();

    public Long getId() {
        return _id;
    }

    public void setId(Long id) {
        _id = id;
    }

    public String getMostId() {
        return _mostId;
    }

    public void setMostId(String mostId) {
        _mostId = mostId;
    }

    public String getName() {
        return _name;
    }

    public void setName(String name) {
        _name = name;
    }

    public String getDescription() {
        return _description;
    }

    public void setDescription(String description) {
        _description = description;
    }

    public String getRelease() {
        return _release;
    }

    public void setRelease(String release) {
        _release = release;
    }

    public boolean isDeleted() {
        return _isDeleted;
    }

    public void setIsDeleted(final boolean deleted) {
        _isDeleted = deleted;
    }

    public Date getDeletedDate() {
        return _deletedDate;
    }

    public void setDeletedDate(final Date deletedDate) {
        _deletedDate = deletedDate;
    }

    public boolean isPermanentlyDeleted() {
        return _isPermanentlyDeleted;
    }

    public void setIsPermanentlyDeleted(final boolean permanentlyDeleted) {
        _isPermanentlyDeleted = permanentlyDeleted;
    }

    public Date getPermanentlyDeletedDate() {
        return _permanentlyDeletedDate;
    }

    public void setPermanentlyDeletedDate(final Date permanentlyDeletedDate) {
        _permanentlyDeletedDate = permanentlyDeletedDate;
    }

    public boolean isApproved() {
        return _isApproved;
    }

    public void setIsApproved(final boolean approved) {
        _isApproved = approved;
    }

    public Long getApprovalReviewId() { return _approvalReviewId; }

    public void setApprovalReviewId(final Long approvalReviewId) { _approvalReviewId = approvalReviewId; }

    public boolean isReleased() {
        return _isReleased;
    }

    public void setIsReleased(boolean released) {
        _isReleased = released;
    }

    public MostFunctionStereotype getFunctionStereotype() {
        return _functionStereotype;
    }

    public void setFunctionStereotype(MostFunctionStereotype functionStereotype) {
        _functionStereotype = functionStereotype;
    }

    public Author getAuthor() {
        return _author;
    }

    public void setAuthor(Author author) {
        _author = author;
    }

    public Company getCompany() {
        return _company;
    }

    public void setCompany(Company company) {
        _company = company;
    }

    public String getReturnParameterName() {
        return _returnParameterName;
    }

    public void setReturnParameterName(final String returnParameterName) {
        _returnParameterName = returnParameterName;
    }

    public String getReturnParameterDescription() {
        return _returnParameterDescription;
    }

    public void setReturnParameterDescription(final String returnParameterDescription) {
        _returnParameterDescription = returnParameterDescription;
    }

    public MostType getReturnType() {
        return _returnType;
    }

    public void setReturnType(MostType returnType) {
        _returnType = returnType;
    }

    public List<Operation> getOperations() {
        return new ArrayList<>(_operations);
    }

    public void addOperation(final Operation operation) {
        _operations.add(operation);
    }

    public void setOperations(final List<Operation> operations) {
        _operations = new ArrayList<>(operations);
    }

    public abstract String getFunctionType();
}

File: www/js/classes/most-function.js
class MostFunction {
    static fromJson(json) {
        const mostFunction = new MostFunction();

        const author = new Author();
        author.setId(json.authorId);
        author.setName(json.authorName);

        const company = new Company();
        company.setId(json.companyId);
        company.setName(json.companyName);

        const operations = [];
        const operationsJson = json.operations;
        for (let i in operationsJson) {
            const operationJson = operationsJson[i];
            const operation = Operation.fromJson(operationJson);
            operations.push(operation);
        }

        const functionType = json.functionType;

        const returnType = new MostType();
        returnType.setId(json.returnTypeId);
        returnType.setName(json.returnTypeName);

        // Only grabbing name and id from JSON. Create and Update Forms will provide all stereotype info from api.
        const mostFunctionStereotype = new MostFunctionStereotype();
        mostFunctionStereotype.setId(json.stereotypeId);
        mostFunctionStereotype.setName(json.stereotypeName);

        mostFunction.setId(json.id);
        mostFunction.setMostId(json.mostId);
        mostFunction.setName(json.name);
        mostFunction.setDescription(json.description);
        mostFunction.setFunctionType(functionType);
        mostFunction.setReleaseVersion(json.releaseVersion);
        mostFunction.setIsReleased(json.isReleased);
        mostFunction.setIsApproved(json.isApproved);
        mostFunction.setApprovalReviewId(json.approvalReviewId);
        mostFunction.setAuthor(author);
        mostFunction.setCompany(company);
        mostFunction.setStereotype(mostFunctionStereotype);
        mostFunction.setReturnParameterName(json.returnParameterName);
        mostFunction.setReturnParameterDescription(json.returnParameterDescription);
        mostFunction.setReturnType(returnType);
        mostFunction.setOperations(operations);
        mostFunction.setIsDeleted(json.isDeleted);
        mostFunction.setDeletedDate(json.deletedDate);

        if (functionType === "Property") {
            mostFunction.setSupportsNotification(json.supportsNotification);
        } else {
            const parameters = [];
            const parametersJson = json.inputParameters;

            for (let i in parametersJson) {
                const parameterJson = parametersJson[i];
                const parameter = Parameter.fromJson(parameterJson);
                parameters.push(parameter);
            }
            mostFunction.setParameters(parameters);
        }

        return mostFunction;
    }

    static toJson(mostFunction) {
        const jsonMostFunction = {
            id:                         mostFunction.getId(),
            mostId:                     formatHex(mostFunction.getMostId()),
            name:                       mostFunction.getName(),
            description:                mostFunction.getDescription(),
            releaseVersion:             mostFunction.getReleaseVersion(),
            isReleased:                 mostFunction.isReleased(),
            isApproved:                 mostFunction.isApproved(),
            functionType:               mostFunction.getFunctionType(),
            supportsNotification:       mostFunction.getSupportsNotification(),
            returnParameterName:        mostFunction.getReturnParameterName(),
            returnParameterDescription: mostFunction.getReturnParameterDescription()
        };

        const returnType = (mostFunction.getReturnType() || new MostType());
        jsonMostFunction.returnTypeId = returnType.getId();
        jsonMostFunction.returnTypeName = returnType.getName();

        const stereotype = (mostFunction.getStereotype() || new MostFunctionStereotype());
        jsonMostFunction.stereotypeId = stereotype.getId();
        jsonMostFunction.stereotypeName = stereotype.getName();

        const author = (mostFunction.getAuthor() || new Author());
        const company = (mostFunction.getCompany() || new Company());
        if (author.getId() > 0) {
            jsonMostFunction.authorId = author.getId();
        }
        if (company.getId() > 0) {
            jsonMostFunction.companyId = company.getId();
        }

        // Jsonify parameters array
        const parameters = mostFunction.getParameters();
        const parametersJson = [];
        for (let i in parameters) {
            const parameterJson = Parameter.toJson(parameters[i]);
            parametersJson.push(parameterJson);
        }
        jsonMostFunction.inputParameters = parametersJson;

        // Jsonify operations array
        const operations = mostFunction.getOperations();
        const operationsJson = [];
        for (let i in operations) {
            const operationJson = Operation.toJson(operations[i]);
            operationsJson.push(operationJson);
        }
        jsonMostFunction.operations = operationsJson;

        return jsonMostFunction;
    }

    constructor() {
      this._id                    = null;
      this._mostId                = "";
      this._name                  = "";
      this._description           = "";
      this._functionType          = "";
      this._releaseVersion        = "";
      this._author                = null;
      this._company               = null;
      this._isReleased            = null;
      this._isApproved            = null;
      this._approvalReviewId      = null;
      this._isDeleted             = false;
      this._deletedDate           = null;

      this._stereotype                  = null;
      this._returnParameterName         = null;
      this._returnParameterDescription  = null;
      this._returnType                  = null;
      this._supportsNotification        = false;

      this._parameters            = [];
      this._operations            = [];
    };

    setId(id) {
        this._id = id;
    }

    getId() {
        return this._id;
    }

    setMostId(mostId) {
        this._mostId = mostId;
    }

    getMostId() {
        return this._mostId;
    }

    setName(name) {
        this._name = name;
    }

    getName() {
        return this._name;
    }

    setDescription(description) {
        this._description = description;
    }

    getDescription() {
        return this._description;
    }

    setFunctionType(functionType) {
        this._functionType = functionType;
    }

    getFunctionType() {
        return this._functionType;
    }

    setReleaseVersion(releaseVersion) {
        this._releaseVersion = releaseVersion;
    }

    getReleaseVersion() {
        return this._releaseVersion;
    }

    setAuthor(author) {
        this._author = author;
    }

    getAuthor() {
        return this._author;
    }

    setCompany(company) {
        this._company = company;
    }

    getCompany() {
        return this._company;
    }

    setStereotype(stereotype) {
        this._stereotype = stereotype;
    }

    getStereotype() {
        return this._stereotype;
    }

    setReturnParameterName(returnParameterName) {
        this._returnParameterName = returnParameterName;
    }

    getReturnParameterName() {
        return this._returnParameterName;
    }

    setReturnParameterDescription(returnParameterDescription) {
        this._returnParameterDescription = returnParameterDescription;
    }

    getReturnParameterDescription() {
        return this._returnParameterDescription;
    }

    setReturnType(returnType) {
        this._returnType = returnType;
    }

    getReturnType() {
        return this._returnType;
    }

    setSupportsNotification(supportsNotification) {
        this._supportsNotification = supportsNotification;
    }

    getSupportsNotification() {
        return this._supportsNotification;
    }

    setParameters(parameters) {
        this._parameters = parameters;
    }

    getParameters() {
        return this._parameters;
    }

    setOperations(operations) {
        this._operations = operations;
    }

    getOperations() {
        return this._operations;
    }

    setIsReleased(isReleased) {
        this._isReleased = isReleased;
    }

    isReleased() {
        return this._isReleased;
    }

    setIsApproved(isApproved) {
        this._isApproved = isApproved;
    }

    isApproved() {
        return this._isApproved;
    }

    setApprovalReviewId(approvalReviewId) {
        this._approvalReviewId = approvalReviewId;
    }

    getApprovalReviewId() {
        return this._approvalReviewId;
    }

    setIsDeleted(isDeleted) {
        this._isDeleted = isDeleted;
    }

    isDeleted() {
        return this._isDeleted;
    }

    setDeletedDate(deletedDate) {
        this._deletedDate = deletedDate;
    }

    getDeletedDate() {
        return this._deletedDate;
    }

}